service/src/main/java/bio/terra/workspace/service/workspace/flight/cloud/aws/CreateWorkspaceApplicationSecurityGroupsStep.java

```diff
@@ -0,0 +1,219 @@
+                .description(
+                    String.format(
+                        "Application Security Group for Workspace %s", workspaceUuid.toString()))
+                .groupName(String.format("ws-app-%s", workspaceUuid.toString()))
+                .tagSpecifications(generateTagSpecification(landingZone, workspaceUuid))
+                .build());
+
+    String groupId = response.groupId();
+    logger.info(
+        "Created EC2 Security Group ID '{}' for Workspace {}", groupId, workspaceUuid.toString());
+    return groupId;
+  }
+
+  /**
+   * Checks whether a security group exists in the AWS region for this workspace. If found, its ID
+   * is returned. If not found, a new security group is created and tagged as belonging to this
+   * workspace.
+   *
+   * @return The ID of the security group that was found or created.
+   */
+  private String findOrCreateSecurityGroupInRegion(
+      AwsCredentialsProvider credentialsProvider,
+      UUID workspaceUuid,
+      Region region,
+      LandingZone landingZone) {
+
+    try (EC2SecurityGroupCow regionCow =
+        EC2SecurityGroupCow.instanceOf(crlService.getClientConfig(), credentialsProvider, region)) {
+      return findSecurityGroupInRegion(regionCow, workspaceUuid)
+          .orElseGet(() -> createSecurityGroupInRegion(regionCow, workspaceUuid, landingZone));
+    }
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+
+    Environment awsEnvironment =
+        awsCloudContextService.discoverEnvironment(
+            FlightUtils.getRequiredUserEmail(flightContext.getInputParameters(), samService));
+
+    AwsCredentialsProvider credentialsProvider =
+        AwsUtils.createWsmCredentialProvider(
+            awsCloudContextService.getRequiredAuthentication(), awsEnvironment);
+
+    Map<String, String> regionSecurityGroups = new HashMap<>();
+
+    // Create a security group in each landing zone.
+    for (Region region : awsEnvironment.getSupportedRegions()) {
+      Optional<LandingZone> landingZone = awsEnvironment.getLandingZone(region);
+
+      // ID's of security groups are not known until after the security group is created.  Thus, in
+      // order to make this step idempotent and not leak security groups, we need to check whether a
+      // security group has already been created for this workspace in a prior invocation by looking
+      // for a security group tagged for this workspace before we actually create a new one.
+
+      String groupId =
+          findOrCreateSecurityGroupInRegion(
+              credentialsProvider, workspaceUuid, region, landingZone.orElseThrow());
+
+      regionSecurityGroups.put(region.toString(), groupId);
+
+      logger.info(
+          "Mapped Security Group ID '{}' to Workspace {} (Landing Zone {})",
+          groupId,
+          workspaceUuid.toString(),
+          region.toString());
+    }
+
+    flightContext
```

Comment:
when i was adding firewall rules in RBS, I updated the map inside the for loop so that if only one region failed, I don't retry on all the other regions too. idk how fast it is for AWS to query for security group.